### ISSUE TITLE
Error when publishing  large message

### ISSUE THREAD

**Issue by luistorres-zz** [moquette-io/moquette#42]
Hello, we are having the following issue after publishing a large message to the broker. The message in question is the application log and can have 1MB. 

Note that the message is published and received by the subscriber clients, but after the following error, we are unable to publish more messages, only after restarting the broker. 

Any further help feel free to ask.

```
Serious error processing the message org.eclipse.moquette.proto.messages.PublishMessage@1731782d for session [clientID: spin-device-16]org.eclipse.moquette.server.netty.NettyChannel@6c2bd990
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
    at org.eclipse.moquette.spi.impl.AnnotationSupport.dispatch(AnnotationSupport.java:66) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.impl.SimpleMessaging.onEvent(SimpleMessaging.java:167) [moquette-broker-0.7-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.impl.SimpleMessaging.onEvent(SimpleMessaging.java:57) [moquette-broker-0.7-SNAPSHOT.jar:na]
    at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:128) [disruptor-3.3.2.jar:na]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_71]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_71]
    at java.lang.Thread.run(Thread.java:745) [na:1.7.0_71]
Caused by: java.lang.reflect.InvocationTargetException: null
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_71]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_71]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_71]
    at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_71]
    at org.eclipse.moquette.spi.impl.AnnotationSupport.dispatch(AnnotationSupport.java:64) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    ... 6 common frames omitted
Caused by: org.mapdb.DBException$VolumeIOError: IO failed
    at org.mapdb.Volume$RandomAccessFileVol.getLong(Volume.java:1896) ~[mapdb-2.0.0-SNAPSHOT.jar:na]
    at org.mapdb.Volume$ReadOnly.getLong(Volume.java:1672) ~[mapdb-2.0.0-SNAPSHOT.jar:na]
    at org.mapdb.StoreDirect.offsetsGet(StoreDirect.java:357) ~[mapdb-2.0.0-SNAPSHOT.jar:na]
    at org.mapdb.StoreDirect.delete2(StoreDirect.java:413) ~[mapdb-2.0.0-SNAPSHOT.jar:na]
    at org.mapdb.StoreCached.flushWriteCacheSegment(StoreCached.java:349) ~[mapdb-2.0.0-SNAPSHOT.jar:na]
    at org.mapdb.StoreWAL.commit(StoreWAL.java:760) ~[mapdb-2.0.0-SNAPSHOT.jar:na]
    at org.mapdb.DB.commit(DB.java:2227) ~[mapdb-2.0.0-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.persistence.MapDBPersistentStore.cleanInFlight(MapDBPersistentStore.java:190) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.impl.ProtocolProcessor.processPublish(ProtocolProcessor.java:309) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.impl.ProtocolProcessor.processPublish(ProtocolProcessor.java:295) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    ... 11 common frames omitted
Caused by: java.io.EOFException: null
    at java.io.RandomAccessFile.readInt(RandomAccessFile.java:827) ~[na:1.7.0_71]
    at java.io.RandomAccessFile.readLong(RandomAccessFile.java:860) ~[na:1.7.0_71]
    at org.mapdb.Volume$RandomAccessFileVol.getLong(Volume.java:1894) ~[mapdb-2.0.0-SNAPSHOT.jar:na]
    ... 20 common frames omitted
```

**Comment by bobwenx** [comment]
seems `MapDB`'s transaction file is incorrect, try to remove `(data_file_name).db.t`(remember to backup) and try start again.

**Comment by luistorres-zz** [comment]
Where can I find that file? we can only find the mapdb and mapdb.wal.

**Comment by andsel** [comment]
The WAL is the transaction log, so I hope it could be deleted.I think that
MapDB 2.0.0 SNAPSHOT has some problem so I switched back to 2.0-alpha in
latest Moquette snapshot, so If you update probably could be solved.

**Comment by luistorres-zz** [comment]
Thank very much for your help, we had already tried to use 2.0-alpha but it didn't fix the issue. Today it worked, maybe the db files were not properly deleted.

**Comment by luistorres-zz** [comment]
The issue still remains, I'm trying to find out what changed.

```
13:18:31.331 o.e.m.s.impl.SimpleMessaging [ERROR] - Serious error processing the message org.eclipse.moquette.proto.messages.PublishMessage@640c7e5c for session [clientID: spin-device-16]org.eclipse.moquette.server.netty.NettyChannel@22c30917
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
    at org.eclipse.moquette.spi.impl.AnnotationSupport.dispatch(AnnotationSupport.java:66) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.impl.SimpleMessaging.onEvent(SimpleMessaging.java:167) [moquette-broker-0.7-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.impl.SimpleMessaging.onEvent(SimpleMessaging.java:57) [moquette-broker-0.7-SNAPSHOT.jar:na]
    at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:128) [disruptor-3.3.2.jar:na]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_71]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_71]
    at java.lang.Thread.run(Thread.java:745) [na:1.7.0_71]
Caused by: java.lang.reflect.InvocationTargetException: null
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_71]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_71]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_71]
    at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_71]
    at org.eclipse.moquette.spi.impl.AnnotationSupport.dispatch(AnnotationSupport.java:64) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    ... 6 common frames omitted
Caused by: org.mapdb.DBException$VolumeIOError: IO failed
    at org.mapdb.Volume$RandomAccessFileVol.getLong(Volume.java:1844) ~[mapdb-2.0-alpha2.jar:na]
    at org.mapdb.Volume$ReadOnly.getLong(Volume.java:1621) ~[mapdb-2.0-alpha2.jar:na]
    at org.mapdb.StoreDirect.offsetsGet(StoreDirect.java:329) ~[mapdb-2.0-alpha2.jar:na]
    at org.mapdb.StoreDirect.delete2(StoreDirect.java:384) ~[mapdb-2.0-alpha2.jar:na]
    at org.mapdb.StoreCached.flushWriteCacheSegment(StoreCached.java:346) ~[mapdb-2.0-alpha2.jar:na]
    at org.mapdb.StoreWAL.commit(StoreWAL.java:753) ~[mapdb-2.0-alpha2.jar:na]
    at org.mapdb.DB.commit(DB.java:2160) ~[mapdb-2.0-alpha2.jar:na]
    at org.eclipse.moquette.spi.persistence.MapDBPersistentStore.cleanInFlight(MapDBPersistentStore.java:190) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.impl.ProtocolProcessor.processPublish(ProtocolProcessor.java:309) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    at org.eclipse.moquette.spi.impl.ProtocolProcessor.processPublish(ProtocolProcessor.java:295) ~[moquette-broker-0.7-SNAPSHOT.jar:na]
    ... 11 common frames omitted
Caused by: java.io.EOFException: null
    at java.io.RandomAccessFile.readInt(RandomAccessFile.java:827) ~[na:1.7.0_71]
    at java.io.RandomAccessFile.readLong(RandomAccessFile.java:860) ~[na:1.7.0_71]
    at org.mapdb.Volume$RandomAccessFileVol.getLong(Volume.java:1842) ~[mapdb-2.0-alpha2.jar:na]
    ... 20 common frames omitted
```

**Comment by andsel** [comment]
Hi @luistorres 
I've tried to reproduce the bug with current Moquette SNAPSHOT, but I wasn't able. I've create a groovy script named [issue42_publishBigMessage.groovy](../blob/master/tools_scripts/integration/issue42_publishBigMessage.groovy) that should reproduce your use case. Please, could you confirm that launching this script on your machine reproduce the bug?

**Comment by luistorres-zz** [comment]
Hi @andsel 
I managed to reproduce the issue by changing the publisher QOS to 1, also the file must be bigger than 65K. 

Any further help feel free to ask.

**Comment by andsel** [comment]
Closed rollbacking MapDB from 2.0.0-SNAPSHOT to 1.0.7
